testing.rst

```diff
@@ -51,79 +50,111 @@ Symfony application.
-in the ``src/Util/`` directory of the app::
+By convention, the ``tests/`` directory should replicate the directory
+of your application for unit tests. So, if you're testing a class in the
+``src/Util/`` directory, put the test in the ``tests/Util/`` directory.
+Autoloading is automatically enabled via the ``vendor/autoload.php`` file
+(as configured by default in the ``phpunit.xml.dist`` file).
 
-    // src/Util/Calculator.php
-    namespace App\Util;
+You can run tests using the ``bin/phpunit`` command:
 
-    class Calculator
-    {
-        public function add($a, $b)
-        {
-            return $a + $b;
-        }
-    }
+.. code-block:: terminal
+
+    # run all tests of the application
+    $ php bin/phpunit
 
-To test this, create a ``CalculatorTest`` file in the ``tests/Util`` directory
-of your application::
+    # run all tests in the Util/ directory
+    $ php bin/phpunit tests/Util
 
-    // tests/Util/CalculatorTest.php
-    namespace App\Tests\Util;
+    # run tests for the Calculator class
+    $ php bin/phpunit tests/Util/CalculatorTest.php
 
-    use App\Util\Calculator;
-    use PHPUnit\Framework\TestCase;
+Integration Tests
+-----------------
 
-    class CalculatorTest extends TestCase
+TODO: KernelTestCase
+
+Accessing the Container
+~~~~~~~~~~~~~~~~~~~~~~~
+
+You can get the same container used in the application, which only includes
+the public services::
+
+    public function testSomething()
     {
-        public function testAdd()
-        {
-            $calculator = new Calculator();
-            $result = $calculator->add(30, 12);
+        $kernel = self::bootKernel();
+        $container = $kernel->getContainer();
+        $someService = $container->get('the-service-ID');
 
-            // assert that your calculator added the numbers correctly!
-            $this->assertEquals(42, $result);
-        }
+        // ...
     }
 
-.. note::
+Symfony tests also have access to a special container that includes both the
+public services and the non-removed :ref:`private services <container-public>`
+services::
 
-    By convention, the ``tests/`` directory should replicate the directory
-    of your application for unit tests. So, if you're testing a class in the
-    ``src/Util/`` directory, put the test in the ``tests/Util/``
-    directory.
+    public function testSomething()
+    {
+        // this call is needed; otherwise the container will be empty
+        self::bootKernel();
 
-Like in your real application - autoloading is automatically enabled via the
-``vendor/autoload.php`` file (as configured by default in the
-``phpunit.xml.dist`` file).
+        $container = self::$container;
+        // $someService = $container->get('the-service-ID');
 
-You can also limit a test run to a directory or a specific test file:
+        // ...
+    }
 
-.. code-block:: terminal
+.. TODO is this really different from self::$container and how to access
+   this in KernelTestCase?
```

Comment:
Excellent question. What about `$kernel->getContainer()->get('test. service_container');`??

Having 3 different ways to get a container is confusing for me. 

Maybe @derrabus knows.

Reply:
Isn't `self::$container` the same object as `$kernel->getContainer()->get('test. service_container');`

Reply:
Yes, and `self::$kernelContainer` is the same as `$kernel->getContainer()`.

However, we also have `$client->getContainer()`, which is defined in `KernelBrowser::getContainer()` as `$kernel->getContainer()`.

In this doc, we somehow make a difference between what's possible with `$client->getContainer()` vs `$kernel->getContainer()` and I'm unsure if they really are different?